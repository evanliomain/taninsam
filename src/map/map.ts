/**
 * @module array=>array
 */
import { Function, Iteratee, Links } from '../@types';
/**
 * This method creates a new array
 * with the results of calling a provided function
 * on every element in the calling array.
 * See the [native map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) for more informations.
 * @param iteratee The iteratee invoked per element.
 * @return the function to apply on the array to transform
 * @example
 * ```
 * map<number, number>(x => 1 + x))([1, 2, 3, 4, 5]) // [2, 3, 4, 5, 6]
 * ```
 * @example Using the chain
 * ```
 * chain([1, 2, 3, 4, 5])
 *   .chain(map<number, number>(x => 1 + x)))
 *   .value() // [2, 3, 4, 5, 6]
 * ```
 */
export function map<T, U>(
  iteratee: Iteratee<T, U>,
): Function<ReadonlyArray<T>, ReadonlyArray<U>> {
  return (array: ReadonlyArray<T>, links?: Links) =>
    array.map((value: T, index: number, originalArray: ReadonlyArray<T>) =>
      iteratee(value, index, originalArray, links),
    );
}
