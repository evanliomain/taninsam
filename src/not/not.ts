/**
 * @module any=>boolean
 */
/**
 * This method return the negation of the boolean result of the passed function,
 * ie: !predicat(x)
 * @param predicat The function that return a boolean.
 * @return the function to apply on the input to return the negation
 * @example
 * ```
 * const isEven = x => 0 === x % 2;
 * const isOdd = not<number>(isEven);
 * isOdd(1) // true
 * ```
 * @example Using the chain
 * ```
 * const isEven = x => 0 === x % 2;
 * chain([1, 2, 3, 4])
 *   .chain(filter(not(isEven))
 *   .value() // [1, 3]
 * ```
 */
export function not<T>(
  predicat: (element: T) => boolean,
): (element: T) => boolean {
  return (element: T) => !predicat(element);
}
