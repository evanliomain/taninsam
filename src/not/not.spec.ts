import { map } from '../map';
import { not } from './not';

describe('not function', () => {
  const input = [1, 2, 3, 4, 5];
  const isEven = (x: number): boolean => 0 === x % 2;

  test('is a pure function', () => {
    map(not(isEven))(input);
    expect(input).toMatchSnapshot();
  });

  test('1 |> not === 1', () => {
    expect(not(isEven)(1)).toBe(true);
  });

  test('x |> not(predicat) calls the predicat', () => {
    const predicat = jest.fn().mockReturnValue(true);
    not(predicat)('foo');
    expect(predicat).toHaveBeenCalledWith('foo');
  });
});
