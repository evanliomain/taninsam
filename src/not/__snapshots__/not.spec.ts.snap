// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`not function is a pure function 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;
